pkg/git/github.go

```diff
@@ -0,0 +1,307 @@
+			Head:                &sourceName,
+			Base:                &targetName,
+			MaintainerCanModify: github.Bool(true),
+		}
+		pr, _, err = g.client.PullRequests.Create(ctx, g.owner, g.repo, createOpts)
+	case 1:
+		pr = (prsOnBranch)[0]
+		pr.Title = &title
+		pr.Body = &description
+		pr.Base.Ref = &targetName
+		pr, _, err = g.client.PullRequests.Edit(ctx, g.owner, g.repo, *pr.Number, pr)
+	default:
+		return fmt.Errorf("received more than one PRs when listing: %d", len(prsOnBranch))
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if auto != (pr.GetAutoMerge() != nil) {
+		client := githubv4.NewClient(g.authClient)
+		var mutation struct {
+			EnablePullRequestAutoMerge struct {
+				PullRequest struct {
+					ID githubv4.ID
+				}
+			} `graphql:"enablePullRequestAutoMerge(input: $input)"`
+		}
+		input := githubv4.EnablePullRequestAutoMergeInput{
+			PullRequestID: pr.GetNodeID(),
+		}
+		err = client.Mutate(ctx, &mutation, input, nil)
+		if err != nil && strings.Contains(err.Error(), "not in the correct state") {
+			err = g.MergePR(ctx, int(pr.GetNumber()), *pr.GetHead().SHA)
+		}
+	}
+	return err
+}
+
+func (g *GitHubGITProvider) GetStatus(ctx context.Context, sha string, group string, env string) (Status, error) {
+	opts := &github.ListOptions{PerPage: 50}
+	statuses, _, err := g.client.Repositories.ListStatuses(ctx, g.owner, g.repo, sha, opts)
+	if err != nil {
+		return Status{}, err
+	}
+
+	name := fmt.Sprintf("%s-%s", group, env)
+	for _, s := range statuses {
+		comp := strings.Split(*s.Context, "/")
+		if len(comp) != 2 {
+			return Status{}, fmt.Errorf("status context in wrong format: %q", *s.Context)
+		}
+		if comp[1] == name {
+			return Status{
+				Succeeded: *s.State == "success",
+			}, nil
+		}
+	}
+
+	return Status{}, fmt.Errorf("no status found for sha %q", sha)
+}
+
+func (g *GitHubGITProvider) SetStatus(ctx context.Context, sha string, group string, env string, succeeded bool) error {
+	description := fmt.Sprintf("%s-%s-%s", group, env, sha)
+	name := fmt.Sprintf("kind/%s-%s", group, env)
+
+	state := "success"
+	if !succeeded {
+		state = "failure"
+	}
+
+	status := &github.RepoStatus{
+		State:       &state,
+		Context:     &name,
+		Description: &description,
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	_, _, err := g.client.Repositories.CreateStatus(ctx, g.owner, g.repo, sha, status)
+	return err
+}
+
+func (g *GitHubGITProvider) MergePR(ctx context.Context, id int, sha string) error {
+	opts := &github.PullRequestOptions{
+		SHA: sha,
+	}
+
+	var result *github.PullRequestMergeResult
+	var res *github.Response
+	err := retry.Do(
+		func() error {
+			var err error
+			result, res, err = g.client.PullRequests.Merge(ctx, g.owner, g.repo, id, "", opts)
+			if err != nil && res.StatusCode == 405 {
+				updateOpts := &github.PullRequestBranchUpdateOptions{}
+				_, _, err = g.client.PullRequests.UpdateBranch(ctx, g.owner, g.repo, id, updateOpts)
```

Comment:
This should always return an error

Reply:
The general provider tests checks for this failure so it seems to be part of the contract.